lib/src/annotate.rs

```diff
@@ -0,0 +1,335 @@
+    /// originator of the line and the string is the actual line itself (without
+    /// newline terminators). The vector is ordered by appearance in the
+    /// file
+    pub file_annotations: Vec<(CommitId, String)>,
+}
+
+struct PartialResults {
+    original_line_map: HashMap<usize, CommitId>,
+    local_line_map: HashMap<CommitId, HashMap<usize, usize>>, /* CommitId -> (local_line_index
+                                                               * -> original_line_index) */
+}
+
+impl PartialResults {
+    fn new(starting_commit_id: &CommitId, num_lines: usize) -> Self {
+        let mut starting_map = HashMap::new();
+        for i in 0..num_lines {
+            starting_map.insert(i, i);
+        }
+        let mut results = PartialResults {
+            original_line_map: HashMap::new(),
+            local_line_map: HashMap::new(),
+        };
+        results
+            .local_line_map
+            .insert(starting_commit_id.clone(), starting_map);
+        results
+    }
+
+    fn forward_to_new_commit(
+        &mut self,
+        old_commit_id: &CommitId,
+        old_local_line_number: usize,
+        new_commit_id: &CommitId,
+        new_local_line_number: usize,
+    ) {
+        let old_map = self.local_line_map.get_mut(old_commit_id).unwrap();
+        let removed_original_line_number = old_map.remove(&old_local_line_number);
+        if removed_original_line_number.is_none() {
+            return;
+        }
+        let removed_original_line_number = removed_original_line_number.unwrap();
+        if self.local_line_map.contains_key(new_commit_id) {
+            self.local_line_map
+                .get_mut(new_commit_id)
+                .unwrap()
+                .insert(new_local_line_number, removed_original_line_number);
+        } else {
+            let mut new_map = HashMap::new();
+            new_map.insert(new_local_line_number, removed_original_line_number);
+            self.local_line_map.insert(new_commit_id.clone(), new_map);
+        }
+    }
+
+    fn swap_full_commit_id(&mut self, old_commit_id: &CommitId, new_commit_id: &CommitId) {
+        let old_commit_map = self.local_line_map.remove(old_commit_id);
+        if old_commit_map.is_none() {
+            return;
+        }
+        let old_commit_map = old_commit_map.unwrap();
+        self.local_line_map
+            .insert(new_commit_id.clone(), old_commit_map);
+    }
+
+    fn drain_remaining_for_commit_id(&mut self, commit_id: &CommitId) {
+        let remaining_lines = self.local_line_map.remove(commit_id);
+        if remaining_lines.is_none() {
+            return;
+        }
+        let remaining_lines = remaining_lines.unwrap();
+        for (_, original_line_number) in remaining_lines {
+            self.original_line_map
+                .insert(original_line_number, commit_id.clone());
+        }
+    }
+
+    fn get_next_commit(&self) -> Option<CommitId> {
+        return self.local_line_map.keys().next().cloned();
```

Comment:
Instead of picking one randomly, I think we should visit parents in topological order, and terminate if there are no more lines to look for.

Mercurial appears to do DFS.

Reply:
I don't think DFS vs BFS matters here. The thing we'll need to guarantee is topological ordering (all descendant lines should have been propagated to the current commit iiuic), and this property is guaranteed if we use revset API. It will also help terminate early.

BTW, `jj annotate lib/src/repo.rs` panics with the current implementation. It might be because of merge commits.

> In terms of graph pruning, I tried to make a few optimizations. Namely, if the parent and child share the same file id, I immediately move on to the parent of the parent

The default index backend wouldn't have any data to optimize it further. (We might want to add bloom filter or something to optimize `log PATH` query.) I assume the Google one will have optimized path for file-oriented query.

Reply:
Nice catch on the bug. Just pushed a fix. 
It was when you have a file with a merge in its history where one side (of the merge) was ignored rather than integrated all changes. 
I've also added a test case to catch this going forward

Reply:
Thanks. Now I understand why random choice works here. Since `forward_to_new_commit()` removes matching lines from the old map, the same matching lines will never propagate to the other parents. So the first parent is always prioritized (= the order is stabilized.)

Still I think topological order is better for optimization. With the current implementation, ancestors of fork point can be visited more than once, where contents are fetched and diffed again.

Reply:
Could we use `RevsetExpression::ancestors()` filtered by path and then use `iter_graph()` to see how the commits are connected? Then it might also work at Google (where iterating over all ancestors is too slow).

Good point about copy tracing. I'll have to think more about that.

Reply:
> Could we use `RevsetExpression::ancestors()` filtered by path and then use `iter_graph()` to see how the commits are connected?

I rejected the idea in my mind, but maybe I was wrong. If both graph walk and line propagation are based on `Vec<GraphEdge>` (instead of `commit.parents()`), it will probably work (and more efficient.)

Reply:
Generally I hesitated about using ancestors() since it most cases, we don't have to check every ancestor so we waste time computing the graph.
In the current implementation, it only goes as deep so as it needs and once it finds changes for every single line, it stops iterating. 
I think where I'm unsatisfied with my implementation is, as you mentioned, some elements of non-determinism due to the `get_next_commit` function. In this case, if two parent commits commit the exact same change to a file, it is not deterministic which commit is taken as the source. This isn't necessarily incorrect, as in that case it is indeed ambiguous which commit is the source commit. However, it would be nice to be able to say which branch is always taken. 
This non-determinism is solved by using variants to ancestors() and whatnot, if, perhaps, a little slower. 

Overall, here is what I'm going to do: based on the comments here, I'm going to try out a few implementations mentioned here and benchmark them across a few different files and cases: Some small files, some large files, some that only need to go shallow in the graph, and some that need to go deep into the commit graph and I'll see how each implementation performs.

I'll try the current the implementation, one with RevsetAncestors, and with RevsetAncestors() filtered to the specific files (of course with iter_graph since it should be lazy).
I'll report back with the results

Reply:
> In this case, if two parent commits commit the exact same change to a file, it is not deterministic which commit is taken as the source. 

Can you use the commit time?

Reply:
> Generally I hesitated about using ancestors() since it most cases, we don't have to check every ancestor so we waste time computing the graph.

Oh, you mean we can omit the second parent if all lines are propagated to the first parent? That's indeed correct. However, I think visiting ancestors (by using the index) is relatively cheap compared to diffing file contents multiple times. And it might avoid abstracting the whole annotation machinery by backends. As Martin said, the current implementation is unusable at Google scale.

`file(PATH)` query is slow with the current default index implementation, but I hope it will be improved later.

> This isn't necessarily incorrect, as in that case it is indeed ambiguous which commit is the source commit. However, it would be nice to be able to say which branch is always taken.

It's also important for test stability.

Reply:
Ok, So I've uploaded a new commit with a revset implementation that uses ancestors + a file filter. It's overall a little faster (maybe 10% faster) when we have to go deeper into the graph. For shallow graphs its a little more expensive but probably not noticeable and not really the cases we are trying to optimize for. 
This new approach is stable in terms of traversing the graph and may be more readable.

@martinvonz how do you think this implementation would perform on large google-like repos? Is there something I'm missing in terms of revsetExpression optimizations?

Overall, what do we think of this approach? I made the changes in a separate commit so we can just choose which one we like better

Reply:
> @martinvonz how do you think this implementation would perform on large google-like repos?

Looks good, thanks!